runtime/commonMain/src/kotlinx/serialization/internal/ObjectSerializer.kt

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2017-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization.internal
+
+import kotlinx.serialization.*
+
+/**
+ * Serializer for Kotlin's singletons (denoted by `object` keyword).
+ *
+ * To preserve singleton instance after serialization and deserialization process, this serializer
+ * accepts instance itself as `theInstance` parameter. This action is automatically performed by a compiler plugin
+ * when you mark `object` as `@Serializable`.
+ *
+ * By default, singleton is serialized as empty structure, which would look in JSON like "{}" (empty object).
+ */
+public class ObjectSerializer<T : Any>(private val serialName: String, private val theInstance: T) : KSerializer<T> {
+    override fun serialize(encoder: Encoder, obj: T) {
+        encoder.beginStructure(descriptor).endStructure(descriptor)
+    }
+
+    override fun deserialize(decoder: Decoder): T {
+        decoder.beginStructure(descriptor).endStructure(descriptor)
+        return theInstance
+    }
+
+    override val descriptor: SerialDescriptor = ObjectDescriptor(serialName)
+}
+
+/**
+ * Object descriptor has UnionKind.OBJECT and consists of itself:
+ * It has one element with name equals to object name
+ * and [getElementDescriptor] returns the descriptor itself.
```

Comment:
This is not a documentation, but rather the description of the implementation

Reply:
I'd suggest applying the same logic as with `EnumDescriptor`: don't document, make internal. Some descriptor traits (such as "It has one element with name equals to object name") will be described in `SerialKind` documentation, because such traits must be true for all descriptors with particular kind, regardless of implementation.